core-api/src/main/java/com/optimizely/ab/odp/ODPEventManager.java

```diff
@@ -0,0 +1,161 @@
+import com.optimizely.ab.odp.serializer.ODPJsonSerializerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.*;
+
+public class ODPEventManager {
+    private static final Logger logger = LoggerFactory.getLogger(ODPEventManager.class);
+    private static final int DEFAULT_BATCH_SIZE = 10;
+    private static final int DEFAULT_QUEUE_SIZE = 10000;
+    private static final int FLUSH_INTERVAL = 1000;
+    private static final int MAX_RETRIES = 3;
+    private static final String EVENT_URL_PATH = "/v3/events";
+
+    private int queueSize = DEFAULT_QUEUE_SIZE;
+    private int batchSize = DEFAULT_BATCH_SIZE;
+
+    private Boolean isRunning = false;
+    private ODPConfig odpConfig;
+    private EventDispatcherThread eventDispatcherThread;
+
+    private final ODPApiManager apiManager;
+    private final List<ODPEvent> currentBatch = new ArrayList<>();
+    private final BlockingQueue<ODPEvent> eventQueue = new LinkedBlockingQueue<>();
+
+    public ODPEventManager(ODPConfig odpConfig, ODPApiManager apiManager) {
+        this.apiManager = apiManager;
+        this.odpConfig = odpConfig;
+    }
+
+    public ODPEventManager(ODPConfig odpConfig, ODPApiManager apiManager, int batchSize, int queueSize) {
+        this(odpConfig, apiManager);
+        this.batchSize = batchSize;
+        this.queueSize = queueSize;
+    }
+
+    public void start() {
+        isRunning = true;
+        eventDispatcherThread = new EventDispatcherThread();
+        eventDispatcherThread.start();
+    }
+
+    public void updateSettings(ODPConfig odpConfig) {
+        this.odpConfig = odpConfig;
+    }
+
+    public void sendEvents(List<ODPEvent> events) {
+        for (ODPEvent event: events) {
+            sendEvent(event);
+        }
+    }
+
+    public void sendEvent(ODPEvent event) {
+        event.setData(augmentCommonData(event.getData()));
+        processEvent(event);
+    }
+
+    private Map<String, Object> augmentCommonData(Map<String, Object> sourceData) {
+        Map<String, Object> data = new HashMap<>();
+        data.put("idempotence_id", UUID.randomUUID().toString());
+        data.put("data_source_type", "sdk");
+        data.put("data_source", ClientEngineInfo.getClientEngine().getClientEngineValue());
+        data.put("data_source_version", BuildVersionInfo.getClientVersion());
+        data.putAll(sourceData);
+        return data;
+    }
+
+    private void processEvent(ODPEvent event) {
+        if (!isRunning) {
+            logger.warn("Failed to Process ODP Event. ODPEventManager is not running");
+            return;
+        }
+
+        if (eventQueue.size() >= queueSize) {
+            logger.warn("Failed to Process ODP Event. Event Queue full. queueSize = " + queueSize);
+            return;
+        }
+
+        if (!eventQueue.offer(event)) {
+            logger.warn("Failed to Process ODP Event. Event Queue is not accepting any more events");
```

Comment:
The usual case would be when queue is at capacity but we are already checking that before this. This check is just there because `offer` can return false in case of an unexpected error. Changing log level to `error`